_posts/2023-10-23-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,312 @@
+    </small></figcaption>
+</figure>
+
+When running Kafka operations, the UTO makes use of the request batching supported by the Kafka Admin client to get higher throughput for metadata operations.
+All `KafkaTopic` events are queued when received, and then processed in batches by a number of controller threads (currently, only a single thread is supported).
+If more than one event related to a single topic resource are found in the batch building, they are put back in the queue to be processed with the next batch.
+
+You can tune the batching mechanism by setting `STRIMZI_MAX_QUEUE_SIZE` (default: 1024), `STRIMZI_MAX_BATCH_SIZE` (default: 100), and `MAX_BATCH_LINGER_MS` (default: 100).
+If you exceed the configured max queue size, the UTO will print an error and then shutdown (Kubernetes will take care of restarting the pod).
+In that case, you can simply raise the max queues size to avoid the periodic operator restarts.
+
+In another UTO test, we set max queue size to `MAX_INT`, max batch size to 200, batch linger ms to 500 and reconciliation interval to 10 seconds.
+With these settings, we were able to achieve a max reconciliation time of 2.5 seconds with 10k concurrent topic events.
+This translates to 3325 new `KafkaTopic` custom resources to become ready in about 2 minutes and 40 seconds.
+
+#### Race condition
+
+There is a known race condition happening when the operator and an external application try to change a topic configuration concurrently in Kafka.
+For example, it could be a GitOps pipeline creating the `KafkaTopic` resource while the application is doing the same during startup.
+
+When the external application wins, the topic will be created using the default Kafka cluster configuration.
+In this case, the UTO tries to reconcile to the desired state, and only fails when an incompatible change is detected (i.e. partition decrease).
+
+The recommendation here is to configure the Kafka cluster with `auto.create.topics.enable: false` in order to reduce the likelihood of hitting this problem.
+Unfortunately, disabling auto topic creation doesn't help with applications that use Kafka Admin client to directly create required topics (i.e. Kafka Streams apps).
+In this case, the application should be written to wait for topic existence, or use an init container to do the same.
+
+#### Finalizers
+
+[Finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers) are used by default to avoid missing topic deletion events when the UTO is not running.
+If this happens, you end up with the `KafkaTopic` resource deleted in Kubernetes by garbage collection, and a Kafka topic still present in Kafka.
+
+It is recommended to delete `KafkaTopic` resources when the operator is running in order to trigger the required cleanups.
+A common pitfall is that the namespace becomes stuck in a "terminating" state when you try to delete it without first deleting all topics.
```

Comment:
Maybe like this?

```suggestion
A common pitfall is that an attempt to delete a namespace or Custom Resource Definition becomes stuck in a "terminating" state when you try to delete everything at once and the operator does not have time to delete all the `KafkaTopic` resources.
```

Reply:
I don't think that deleting CRDs is that common.

Reply:
I think users who:
* Install the operator with `kubectl apply -f ...`
* Do something
* Try to uninstall the operator with `kubectl delete -f ...` on the same set of installation files and get stuck will be at least as common as users deleting the namespace.

Reply:
That's true. Thanks for raising.